### PR TITLE
Calculate exponent of predictions for Poisson regression

### PR DESCRIPTION
**Pull request by garkavem**
Before submitting a pull request, please do the following steps:

1. Read [instructions for contributors](https://catboost.ai/docs/concepts/development-and-contributions.html).
2. Run `ya make` in catboost folder to make sure the code builds.
3. Add tests that test your change.
4. Run tests using `ya make -t -A` command.
5. If you haven't already, complete the [CLA](https://yandex.ru/legal/cla/).

### CODE REVIEW

File: catboost/libs/model/eval_processing.h [hunk 1]
```diff
@@ -114,22 +123,22 @@ namespace NCB::NModelEvaluation {
...
-                blockId * BlockSize * dimension,
-                Min<ui32>(
-                    BlockSize * dimension,
-                    Results.size() - (blockId * BlockSize * dimension)
-                )
+                    blockId * BlockSize * dimension,
+                    Min<ui32>(
+                            BlockSize * dimension,
+                            Results.size() - (blockId * BlockSize * dimension)
+                    )
```
Comment (nikitxskv) [inline catboost/libs/model/eval_processing.h]: Return old indents please)
And in other places too.

File: catboost/libs/model/eval_processing.h [hunk 2]
```diff
@@ -98,6 +98,15 @@ inline void CalcLogSigmoid(const TConstArrayRef<double> approx, TArrayRef<double
...
 
+inline TVector<double> CalcExponent(const TConstArrayRef<double> approx) {
+    TVector<double> exponents;
+    exponents.yresize(approx.size());
+    for (size_t i = 0; i < approx.size(); ++i) {
+        exponents[i] = exp(approx[i]);
+    }
+    return exponents;
+}
+
```
Comment (nikitxskv) [inline catboost/libs/model/eval_processing.h]: Why is this function not used?

File: catboost/libs/eval_result/eval_helpers.cpp [hunk 3]
```diff
@@ -217,6 +217,13 @@ void PrepareEval(const EPredictionType predictionType,
                 }
             }
             break;
+        case EPredictionType::Poisson:
+            result->resize(1);
+            (*result)[0].reserve(approx.size());
+            for (const double prediction : approx[0]) {
+                (*result)[0].push_back(exp(prediction));
+            }
```
Comment (nikitxskv) [inline catboost/libs/eval_result/eval_helpers.cpp]: It's better to use [this](https://github.com/catboost/catboost/blob/master/catboost/libs/model/eval_processing.h#L15) function here. It is much faster.

File: catboost/libs/model/eval_processing.h (hunk 1 above)
Comment (garkavem) [reply catboost/libs/model/eval_processing.h]: Ok (actually CLion changed them without my consent )

File: catboost/libs/model/eval_processing.h [hunk 4]
```diff
@@ -27,6 +27,19 @@ inline void CalcSoftmax(const TConstArrayRef<double> approx, TArrayRef<double> s
     }
 }
 
+inline TVector<double> CalcExponent(const TConstArrayRef<double> approx) {
+    TVector<double> exponents;
+    exponents.yresize(approx.size());
+
+    for (size_t dim = 0; dim < approx.size(); ++dim) {
+        exponents[dim] = approx[dim];
+    }
```
Comment (nikitxskv) [inline catboost/libs/model/eval_processing.h]: It's better to take this `approx` argument by value. In this case you don't need to manual copy it:
```
inline TVector<double> CalcExponent(TVector<double> approx) {
    FastExpInplace(approx.data(), approx.ysize());
    return approx;
}
```

File: catboost/libs/eval_result/eval_helpers.cpp [hunk 5]
```diff
@@ -217,6 +217,9 @@ void PrepareEval(const EPredictionType predictionType,
                 }
             }
             break;
+        case EPredictionType::Exponent:
+            *result = {CalcExponent(approx[0])};
+            break;
```
Comment (nikitxskv) [inline catboost/libs/eval_result/eval_helpers.cpp]: Let's add EPredictionType::Exponent in this place too:
https://github.com/catboost/catboost/blob/master/catboost/libs/model/eval_processing.h#L158-L165

File: catboost/python-package/catboost/core.py [hunk 6]
```diff
@@ -1835,8 +1835,8 @@ def _process_predict_input_data(self, data, parent_method_name, thread_count, la
     def _validate_prediction_type(self, prediction_type):
         if not isinstance(prediction_type, STRING_TYPES):
             raise CatBoostError("Invalid prediction_type type={}: must be str().".format(type(prediction_type)))
-        if prediction_type not in ('Class', 'RawFormulaVal', 'Probability', 'LogProbability'):
-            raise CatBoostError("Invalid value of prediction_type={}: must be Class, RawFormulaVal, Probability, LogProbability.".format(prediction_type))
+        if prediction_type not in ('Class', 'RawFormulaVal', 'Probability', 'LogProbability', 'Exponent'):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py]: Let's add test.
Example: https://github.com/catboost/catboost/blob/master/catboost/python-package/ut/medium/test.py#L7048-L7056

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@garkavem, please agree to the CLA (https://github.com/catboost/catboost/blob/master/CONTRIBUTING.md#provide-contributions)

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [1195509](https://a.yandex-team.ru/review/1195509)

**Comment by arcadia-devtools** [comment]
Check [1195509](https://a.yandex-team.ru/review/1195509) failed: 33 test(s) broken

**Comment by garkavem** [comment]
I hereby agree to the terms of the CLA available at:  https://yandex.ru/legal/cla/?lang=en

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [1195605](https://a.yandex-team.ru/review/1195605)

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
Check [1195605](https://a.yandex-team.ru/review/1195605) failed: 7 test(s) broken

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [1195670](https://a.yandex-team.ru/review/1195670)

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
Check [1195670](https://a.yandex-team.ru/review/1195670) passed

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [1195718](https://a.yandex-team.ru/review/1195718)

**Pull request merged**